app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/SpatialCache.kt

```diff
@@ -0,0 +1,138 @@
+package de.westnordost.streetcomplete.data.osm.osmquests
+
+import android.util.Log
+import de.westnordost.streetcomplete.data.download.tiles.TilePos
+import de.westnordost.streetcomplete.data.download.tiles.enclosingTilePos
+import de.westnordost.streetcomplete.data.download.tiles.enclosingTilesRect
+import de.westnordost.streetcomplete.data.download.tiles.minTileRect
+import de.westnordost.streetcomplete.data.osm.mapdata.BoundingBox
+import de.westnordost.streetcomplete.data.osm.mapdata.LatLon
+import de.westnordost.streetcomplete.util.math.contains
+import de.westnordost.streetcomplete.util.math.isCompletelyInside
+
+// inspired by Android LruCache
+// fetch needs get data from db and fill other caches, like questKey -> Quest
+// onRemoveTile needs to clean other caches, like questKey -> Quest
+// careful: put does not actually put key into cache if tile is not cached -> update db blocking if key was not put
+// TODO: synchronized stuff? didn't get related crash so far...
```

Comment:
SpatialCache is never used alone but always in combination with some `HashMap<Long, Quest>` etc., so not this class needs to be threadsafe but the code using SpatialCache + HashMap(s)